Improvements

pepeiborra opened the pull request:
Last week at the office @ndmitchell was looking for a debugging tool and I tried to demo the new version of Hoed. Unfortunately it didn't work very well, we couldn't observe even a simple example. 

These changes are a first stab at improving the first experience with Hoed, the main bits are:
- Fixed a couple of odd bugs (head of empty list, empty regex, indexing errors, incomplete pattern matches)
- Fixed regexes in Windows
- Re-added the TH primitive that I contributed to the previous version of Hoed
- Added a new command to show the list of available observations 
- Made a couple of minor performance fixes (avoid nub and use atomicModify instead of locks)

There is still a lot of work to do: 
- The documentation and the haddocks still refer to the GUI instead of the new CLI. 
- The CLI itself needs a bit of love to become as usable as the GUI.
- The original Hood had the ability to do nested observations, but Hoed removed that.

Apologies for the large number of changes. Let me know if there is any way I can make them easier to review.

MaartenFaddegon commented:
Hi Pepe

Thank you for the many valuable contributions. I'll review your changes between Christmas and New Years, and, time permitting, I will also try to address your other remarks.

Could you maybe give an example of a nested observation?

pepeiborra commented:
Thanks Maarten,

I will raise issues for my other remarks so that we can track them and avoid duplicate work.

Nested observations are described in section 4.4 of the Hood paper, but it occurs to me that perhaps in Hoed one does not need the explicit nesting operator, as according to your PLDI paper the hierarchy of dependency should be already in the Computational Tree.

pepeiborra commented:
I've updated the pull request to include nested observations, extracted from the computational tree. This works by having the observe command to also print the first level of descendants comp statements which correspond to constants. 
It looks like this:

```
hdb> observe 
--- stmt-1 ------------------------------------------
quickSort2 "ae"
  
  = "ae"
  where
    lt = []
    gt = "e"
--- stmt-2 ------------------------------------------
quickSort2 "e"
  
  = "e"
  where
    lt = []
    gt = []
--- stmt-3 ------------------------------------------
quickSort2 "haskell"
  
  = "aehklls"
  where
    lt = "ae"
    gt = "skll"
```

MaartenFaddegon commented:
Would it be possible to split this into two separate pull requests?

Your original changes seem quit straightforward improvements and I am happy to merge these very soon.

I have some concerns about soundness for algorithmic debugging of your work on nested observations, and I need a bit more time to study what it exactly does. In Chapter 7 of my phd thesis I discuss these concerns in some more detail (http://maartenfaddegon.nl/#pub). 

Travis also indicates that not all versions of GHC can compile the code you later added to the pull request.

pepeiborra commented:
The GHC < 7.10 issues have been fixed in commit 9c45b11. 

Although I've only skimmed through your thesis (including chapter 7), I read the PLDI paper in full. The debug TH command adds observations to bindings defined in where clauses; correct me if I'm wrong, but as long as these bindings are constants, then soundness is preserved. If we wanted, it would be trivial to handle lambdas with no free variables as well, although I am happy with just constants for now.

Let me know if you still have concerns and I will split the TH command out to a separate package.

pepeiborra commented:
These changes have enabled the development of [debug-hoed](https://github.com/pepeiborra/debug-hoed). See [an example](https://rawgit.com/pepeiborra/debug-hoed/master/example/quicksort.html) of a debugging session with debug-hoed.

pepeiborra commented:
Awesome, thanks !

pepeiborra commented:
Any chance of a Hackage release this week ?

MaartenFaddegon commented:
Would that help you? Because then I can do, but otherwise I wait until I have more time to dig up some of the more "tricky" example programs from my phd research to convince myself of soundness for algorithmic debugging with the new annotations.

pepeiborra commented:
No worries, take your time. I want to release `debug-hoed`, but I can do that with an embedded snapshot of `Hoed`, and turn that into a dependency once you have a chance to review in depth.
